File: services/brand-rules/src/lumi_brand_rules/runtime.py
from __future__ import annotations

from typing import Any, Mapping

from .model import (
    BrandAssetSet,
    BrandComplianceReport,
    BrandDiagnostic,
    BrandRule,
    BrandRuleError,
    BrandRuleSet,
    BrandTokenSet,
    validate_rule_set,
)


def _value(node: Mapping[str, Any], *paths: str) -> Any:
    for path in paths:
        current: Any = node
        for key in path.split("."):
            if not isinstance(current, Mapping):
                current = None
                break
            current = current.get(key)
        if current is not None:
            return current
    return None


def _string_list(value: Any) -> tuple[str, ...]:
    if not isinstance(value, (list, tuple)):
        return ()
    return tuple(item for item in value if isinstance(item, str))


def _applies(rule: BrandRule, node: Mapping[str, Any], channel: str | None, locale: str | None) -> bool:
    node_ids = _string_list(rule.scope.get("node_ids"))
    roles = _string_list(rule.scope.get("roles"))
    channels = _string_list(rule.scope.get("channels"))
    locales = _string_list(rule.scope.get("locales"))
    if node_ids and node.get("id") not in node_ids:
        return False
    if roles and node.get("role") not in roles:
        return False
    if channels and channel not in channels:
        return False
    return not locales or locale in locales


def _repair(document: Mapping[str, Any], node_id: str, path: str, value: Any, rule_id: str) -> Mapping[str, Any]:
    metadata = document.get("metadata", {})
    version = metadata.get("document_version", 0) if isinstance(metadata, Mapping) else 0
    return {
        "operation_id": f"brand-fix:{rule_id}:{node_id}:{path}",
        "type": "SET_PROPERTY",
        "target_ids": [node_id],
        "expected_document_version": version if isinstance(version, int) else 0,
        "payload": {"path": path, "value": value},
        "reason": f"BRAND_RULE_AUTO_FIX:{rule_id}",
    }


def _diag(
    rule: BrandRule,
    reason: str,
    node_id: str | None = None,
    expected: Any = None,
    actual: Any = None,
    repairs: tuple[Mapping[str, Any], ...] = (),
) -> BrandDiagnostic:
    return BrandDiagnostic(
        rule_id=rule.id,
        severity=rule.severity,
        category=rule.category,
        reason_code=reason,
        node_id=node_id,
        expected=expected,
        actual=actual,
        repair_operations=repairs,
    )


def _rect(node: Mapping[str, Any]) -> tuple[float, float, float, float] | None:
    transform = node.get("transform")
    if not isinstance(transform, Mapping):
        return None
    values = tuple(transform.get(key) for key in ("x", "y", "width", "height"))
    if not all(isinstance(value, (int, float)) for value in values):
        return None
    return tuple(float(value) for value in values)  # type: ignore[arg-type,return-value]


def _overlap_margin(
    a: tuple[float, float, float, float],
    b: tuple[float, float, float, float],
    margin: float,
) -> bool:
    ax, ay, aw, ah = a
    bx, by, bw, bh = b
    return not (
        bx >= ax + aw + margin
        or bx + bw <= ax - margin
        or by >= ay + ah + margin
        or by + bh <= ay - margin
    )


def _evaluate_rule(
    rule: BrandRule,
    document: Mapping[str, Any],
    token_set: BrandTokenSet,
    verified_asset_ids: frozenset[str] | None,
    font_rights_ids: frozenset[str] | None,
    channel: str | None,
    locale: str | None,
) -> list[BrandDiagnostic]:
    if not rule.active:
        return []
    nodes_raw = document.get("nodes", {})
    if not isinstance(nodes_raw, Mapping):
        return []
    nodes = [node for node in nodes_raw.values() if isinstance(node, Mapping)]
    nodes.sort(key=lambda item: str(item.get("id", "")))
    diagnostics: list[BrandDiagnostic] = []

    for node in nodes:
        if not _applies(rule, node, channel, locale):
            continue
        node_id = str(node.get("id", ""))
        if rule.type == "FORBIDDEN_COLORS":
            color = _value(node, "fill", "color", "style.fill", "metadata.fill")
            forbidden = {item.lower() for item in _string_list(rule.parameters.get("colors"))}
            if isinstance(color, str) and color.lower() in forbidden:
                replacement = next(iter(token_set.colors.values()), None)
                repairs = (
                    _repair(document, node_id, "fill", replacement, rule.id),
                ) if replacement else ()
                diagnostics.append(_diag(rule, "BRAND_COLOR_FORBIDDEN", node_id, sorted(forbidden), color, repairs))

        elif rule.type == "ALLOWED_FONT_ASSETS" and node.get("kind") == "TEXT":
            font_id = _value(node, "font_asset_id", "typography.font_asset_id", "metadata.font_asset_id")
            allowed = _string_list(rule.parameters.get("asset_ids"))
            rights_denied = isinstance(font_id, str) and font_rights_ids is not None and font_id not in font_rights_ids
            if not isinstance(font_id, str) or font_id not in allowed or rights_denied:
                replacement = next((item for item in allowed if font_rights_ids is None or item in font_rights_ids), None)
                repairs = (
                    _repair(document, node_id, "font_asset_id", replacement, rule.id),
                ) if replacement else ()
                diagnostics.append(_diag(
                    rule,
                    "BRAND_FONT_RIGHTS_UNAVAILABLE" if rights_denied else "BRAND_FONT_NOT_ALLOWED",
                    node_id,
                    allowed,
                    font_id,
                    repairs,
                ))

        elif rule.type == "MIN_TEXT_SIZE" and node.get("kind") == "TEXT":
            size = _value(node, "font_size", "typography.font_size", "metadata.font_size")
            minimum = rule.parameters.get("px", 0)
            if isinstance(size, (int, float)) and isinstance(minimum, (int, float)) and size < minimum:
                diagnostics.append(_diag(
                    rule,
                    "BRAND_TEXT_TOO_SMALL",
                    node_id,
                    minimum,
                    size,
                    (_repair(document, node_id, "font_size", minimum, rule.id),),
                ))

        elif rule.type == "REQUIRE_TOKEN_BINDING":
            binding = _value(node, "brand_binding", "metadata.brand_binding")
            prefixes = _string_list(rule.parameters.get("prefixes"))
            valid = isinstance(binding, str) and (not prefixes or any(binding.startswith(item) for item in prefixes))
            if not valid:
                repairs = (
                    _repair(document, node_id, "brand_binding", prefixes[0], rule.id),
                ) if prefixes else ()
                diagnostics.append(_diag(rule, "BRAND_TOKEN_BINDING_REQUIRED", node_id, prefixes, binding, repairs))

        elif rule.type == "LOGO_FORBID_ROTATION":
            rotation = _value(node, "transform.rotation_deg") or 0
            tolerance = rule.parameters.get("tolerance_deg", 0.01)
            if isinstance(rotation, (int, float)) and isinstance(tolerance, (int, float)) and abs(rotation) > tolerance:
                diagnostics.append(_diag(
                    rule,
                    "BRAND_LOGO_ROTATED",
                    node_id,
                    0,
                    rotation,
                    (_repair(document, node_id, "transform.rotation_deg", 0, rule.id),),
                ))

        elif rule.type == "LOGO_CLEAR_SPACE":
            bounds = _rect(node)
            margin = rule.parameters.get("px", 0)
            if bounds and isinstance(margin, (int, float)) and margin > 0:
                for other in nodes:
                    if other is node or other.get("visible") is False or other.get("kind") == "GUIDE":
                        continue
                    other_bounds = _rect(other)
                    if other_bounds and _overlap_margin(bounds, other_bounds, float(margin)):
                        diagnostics.append(_diag(
                            rule,
                            "BRAND_LOGO_CLEAR_SPACE_VIOLATION",
                            node_id,
                            margin,
                            other.get("id"),
                        ))
                        break

        elif rule.type in {"ALLOWED_ASSETS", "ALLOWED_LOGO_ASSETS"}:
            asset_id = _value(node, "asset_id", "resource_id", "metadata.asset_id")
            allowed = _string_list(rule.parameters.get("asset_ids"))
            if not isinstance(asset_id, str) or asset_id not in allowed:
                diagnostics.append(_diag(rule, "BRAND_ASSET_NOT_ALLOWED", node_id, allowed, asset_id))
            elif verified_asset_ids is not None and asset_id not in verified_asset_ids:
                diagnostics.append(_diag(rule, "BRAND_ASSET_NOT_VERIFIED", node_id, True, asset_id))

        elif rule.type == "VOICE_FORBIDDEN_TERMS" and node.get("kind") == "TEXT":
            text = _value(node, "text", "content", "metadata.text")
            terms = _string_list(rule.parameters.get("terms"))
            if isinstance(text, str):
                hit = next((term for term in terms if term.lower() in text.lower()), None)
                if hit:
                    diagnostics.append(_diag(rule, "BRAND_VOICE_FORBIDDEN_TERM", node_id, terms, hit))

    return diagnostics


def evaluate_brand_compliance(
    document: Mapping[str, Any],
    rule_set: BrandRuleSet,
    token_set: BrandTokenSet,
    asset_set: BrandAssetSet,
    *,
    verified_asset_ids: frozenset[str] | None = None,
    font_rights_allowed_asset_ids: frozenset[str] | None = None,
    channel: str | None = None,
    locale: str | None = None,
) -> BrandComplianceReport:
    validate_rule_set(rule_set)
    if rule_set.status != "PUBLISHED":
        raise BrandRuleError("compliance requires a PUBLISHED BrandRuleSet")
    if token_set.version != rule_set.token_set_version or asset_set.version != rule_set.asset_set_version:
        raise BrandRuleError("brand dependency version mismatch")
    if token_set.brand_profile_id != rule_set.brand_profile_id or asset_set.brand_profile_id != rule_set.brand_profile_id:
        raise BrandRuleError("brand profile mismatch")

    diagnostics: list[BrandDiagnostic] = []
    for rule in sorted(rule_set.rules, key=lambda item: (-item.priority, item.id)):
        diagnostics.extend(_evaluate_rule(
            rule,
            document,
            token_set,
            verified_asset_ids,
            font_rights_allowed_asset_ids,
            channel,
            locale,
        ))
    hard = sum(item.severity == "HARD" for item in diagnostics)
    soft = sum(item.severity == "SOFT" for item in diagnostics)
    advisory = sum(item.severity == "ADVISORY" for item in diagnostics)
    decision = "FAIL" if hard else "PASS_WITH_WARNINGS" if diagnostics else "PASS"
    return BrandComplianceReport(
        brand_rule_set_version=rule_set.version,
        decision=decision,
        score=max(0.0, min(1.0, 1 - hard * 0.25 - soft * 0.08 - advisory * 0.02)),
        diagnostics=tuple(diagnostics),
        hard_violation_count=hard,
        soft_violation_count=soft,
        advisory_count=advisory,
    )

File: packages/brand-rules/src/runtime.ts
import { getDocumentVersion } from "../../design-ir/src/index";
import type { DesignNode, DesignOperation, JsonValue } from "../../design-ir/src/index";
import type {
  BrandAssetSet,
  BrandComplianceReport,
  BrandDiagnostic,
  BrandEvaluationContext,
  BrandRule,
  BrandRuleSet,
  BrandTokenSet,
} from "./types";

export class BrandRuleError extends Error {}

function stringArray(value: JsonValue | undefined): readonly string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string") : [];
}

function numberValue(value: JsonValue | undefined, fallback: number): number {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

function rawNode(node: DesignNode): Readonly<Record<string, unknown>> {
  return node as unknown as Readonly<Record<string, unknown>>;
}

function nested(node: DesignNode, path: string): unknown {
  let current: unknown = rawNode(node);
  for (const key of path.split(".")) {
    if (current === null || typeof current !== "object" || Array.isArray(current)) return undefined;
    current = (current as Readonly<Record<string, unknown>>)[key];
  }
  return current;
}

function firstString(node: DesignNode, paths: readonly string[]): string | undefined {
  for (const path of paths) {
    const value = nested(node, path);
    if (typeof value === "string" && value) return value;
  }
  return undefined;
}

function normalizedColor(value: string): string {
  return value.trim().toLowerCase();
}

function nodeColor(node: DesignNode): string | undefined {
  return firstString(node, ["fill", "color", "background_color", "style.fill", "metadata.fill"]);
}

function nodeAssetId(node: DesignNode): string | undefined {
  return firstString(node, ["asset_id", "resource_id", "metadata.asset_id", "metadata.resource_id"]);
}

function nodeFontAssetId(node: DesignNode): string | undefined {
  return firstString(node, ["font_asset_id", "typography.font_asset_id", "metadata.font_asset_id"]);
}

function nodeText(node: DesignNode): string | undefined {
  return firstString(node, ["text", "content", "metadata.text"]);
}

function nodeBrandBinding(node: DesignNode): string | undefined {
  return firstString(node, ["brand_binding", "metadata.brand_binding"]);
}

function appliesToRule(rule: BrandRule, node: DesignNode, context: BrandEvaluationContext): boolean {
  const { scope } = rule;
  if (scope.node_ids?.length && !scope.node_ids.includes(node.id)) return false;
  if (scope.roles?.length && (!node.role || !scope.roles.includes(node.role))) return false;
  if (scope.channels?.length && (!context.channel || !scope.channels.includes(context.channel))) return false;
  if (scope.locales?.length && (!context.locale || !scope.locales.includes(context.locale))) return false;
  return true;
}

function repairOperation(
  context: BrandEvaluationContext,
  nodeId: string,
  path: string,
  value: unknown,
  ruleId: string,
): DesignOperation {
  return {
    operation_id: `brand-fix:${ruleId}:${nodeId}:${path}`,
    type: "SET_PROPERTY",
    target_ids: [nodeId],
    expected_document_version: getDocumentVersion(context.document),
    payload: { path, value },
    reason: `BRAND_RULE_AUTO_FIX:${ruleId}`,
  };
}

function diagnostic(
  rule: BrandRule,
  reasonCode: string,
  nodeId?: string,
  expected?: JsonValue,
  actual?: JsonValue,
  repairOperations?: readonly DesignOperation[],
): BrandDiagnostic {
  return {
    rule_id: rule.id,
    severity: rule.severity,
    category: rule.category,
    reason_code: reasonCode,
    ...(nodeId ? { node_id: nodeId } : {}),
    ...(expected !== undefined ? { expected } : {}),
    ...(actual !== undefined ? { actual } : {}),
    ...(repairOperations?.length ? { repair_operations: repairOperations } : {}),
  };
}

function tokenColorMap(tokenSet: BrandTokenSet): Map<string, string> {
  return new Map(tokenSet.colors.map((token) => [token.id, normalizedColor(token.value)]));
}

function evaluateColor(rule: BrandRule, node: DesignNode, context: BrandEvaluationContext): BrandDiagnostic[] {
  const color = nodeColor(node);
  if (!color) return [];
  const normalized = normalizedColor(color);
  const tokenColors = tokenColorMap(context.token_set);

  if (rule.type === "ALLOWED_COLOR_TOKENS") {
    const tokenIds = stringArray(rule.parameters.token_ids);
    const allowed = new Set(tokenIds.map((id) => tokenColors.get(id)).filter((item): item is string => Boolean(item)));
    if (!allowed.size || allowed.has(normalized)) return [];
    const replacementToken = tokenIds.find((id) => tokenColors.has(id));
    const replacement = replacementToken ? tokenColors.get(replacementToken) : undefined;
    return [diagnostic(
      rule,
      "BRAND_COLOR_NOT_ALLOWED",
      node.id,
      tokenIds,
      color,
      replacement ? [repairOperation(context, node.id, "fill", replacement, rule.id)] : undefined,
    )];
  }

  if (rule.type === "FORBIDDEN_COLORS") {
    const forbidden = new Set(stringArray(rule.parameters.colors).map(normalizedColor));
    if (!forbidden.has(normalized)) return [];
    const replacement = context.token_set.colors[0]?.value;
    return [diagnostic(
      rule,
      "BRAND_COLOR_FORBIDDEN",
      node.id,
      stringArray(rule.parameters.colors),
      color,
      replacement ? [repairOperation(context, node.id, "fill", replacement, rule.id)] : undefined,
    )];
  }
  return [];
}

function evaluateTypography(rule: BrandRule, node: DesignNode, context: BrandEvaluationContext): BrandDiagnostic[] {
  if (node.kind !== "TEXT") return [];
  if (rule.type === "ALLOWED_FONT_ASSETS") {
    const assetId = nodeFontAssetId(node);
    const allowed = new Set(stringArray(rule.parameters.asset_ids));
    if (!assetId || !allowed.has(assetId)) {
      const replacement = [...allowed].find((id) => context.font_rights_allowed_asset_ids?.includes(id) ?? true);
      return [diagnostic(
        rule,
        assetId && !context.font_rights_allowed_asset_ids?.includes(assetId)
          ? "BRAND_FONT_RIGHTS_UNAVAILABLE"
          : "BRAND_FONT_NOT_ALLOWED",
        node.id,
        [...allowed],
        assetId ?? null,
        replacement ? [repairOperation(context, node.id, "font_asset_id", replacement, rule.id)] : undefined,
      )];
    }
    if (context.font_rights_allowed_asset_ids && !context.font_rights_allowed_asset_ids.includes(assetId)) {
      return [diagnostic(rule, "BRAND_FONT_RIGHTS_UNAVAILABLE", node.id, [...allowed], assetId)];
    }
  }

  if (rule.type === "MIN_TEXT_SIZE") {
    const size = nested(node, "font_size") ?? nested(node, "typography.font_size") ?? nested(node, "metadata.font_size");
    const minimum = numberValue(rule.parameters.px, 0);
    if (typeof size === "number" && size < minimum) {
      return [diagnostic(
        rule,
        "BRAND_TEXT_TOO_SMALL",
        node.id,
        minimum,
        size,
        [repairOperation(context, node.id, "font_size", minimum, rule.id)],
      )];
    }
  }
  return [];
}

function rect(node: DesignNode): { x: number; y: number; width: number; height: number } | null {
  const transform = node.transform;
  if (!transform) return null;
  const { x, y, width, height } = transform;
  if (![x, y, width, height].every((value) => typeof value === "number" && Number.isFinite(value))) return null;
  return { x: x as number, y: y as number, width: width as number, height: height as number };
}

function overlapsWithMargin(
  a: { x: number; y: number; width: number; height: number },
  b: { x: number; y: number; width: number; height: number },
  margin: number,
): boolean {
  return !(
    b.x >= a.x + a.width + margin ||
    b.x + b.width <= a.x - margin ||
    b.y >= a.y + a.height + margin ||
    b.y + b.height <= a.y - margin
  );
}

function evaluateLogo(rule: BrandRule, node: DesignNode, context: BrandEvaluationContext): BrandDiagnostic[] {
  const diagnostics: BrandDiagnostic[] = [];
  const assetId = nodeAssetId(node);
  if (rule.type === "ALLOWED_LOGO_ASSETS") {
    const allowed = new Set(stringArray(rule.parameters.asset_ids));
    if (!assetId || !allowed.has(assetId)) {
      diagnostics.push(diagnostic(rule, "BRAND_LOGO_ASSET_NOT_ALLOWED", node.id, [...allowed], assetId ?? null));
    }
  }
  if (rule.type === "LOGO_MIN_SIZE") {
    const bounds = rect(node);
    if (bounds) {
      const minWidth = numberValue(rule.parameters.min_width, 0);
      const minHeight = numberValue(rule.parameters.min_height, 0);
      if (bounds.width < minWidth || bounds.height < minHeight) {
        diagnostics.push(diagnostic(rule, "BRAND_LOGO_TOO_SMALL", node.id, { min_width: minWidth, min_height: minHeight }, { width: bounds.width, height: bounds.height }));
      }
    }
  }
  if (rule.type === "LOGO_FORBID_ROTATION") {
    const rotation = node.transform?.rotation_deg ?? 0;
    const tolerance = numberValue(rule.parameters.tolerance_deg, 0.01);
    if (Math.abs(rotation) > tolerance) {
      diagnostics.push(diagnostic(rule, "BRAND_LOGO_ROTATED", node.id, 0, rotation, [repairOperation(context, node.id, "transform.rotation_deg", 0, rule.id)]));
    }
  }
  if (rule.type === "LOGO_FORBID_STRETCH") {
    const sx = node.transform?.scale_x ?? 1;
    const sy = node.transform?.scale_y ?? 1;
    const tolerance = numberValue(rule.parameters.tolerance, 0.01);
    if (Math.abs(sx - sy) > tolerance) diagnostics.push(diagnostic(rule, "BRAND_LOGO_STRETCHED", node.id, "uniform-scale", { scale_x: sx, scale_y: sy }));
  }
  if (rule.type === "LOGO_FORBID_RECOLOR") {
    const recolored = nested(node, "metadata.logo_recolored");
    if (recolored === true) diagnostics.push(diagnostic(rule, "BRAND_LOGO_RECOLORED", node.id, false, true));
  }
  if (rule.type === "LOGO_CLEAR_SPACE") {
    const bounds = rect(node);
    const margin = numberValue(rule.parameters.px, 0);
    if (bounds && margin > 0) {
      for (const other of Object.values(context.document.nodes)) {
        if (other.id === node.id || other.visible === false || other.kind === "GUIDE") continue;
        const otherBounds = rect(other);
        if (otherBounds && overlapsWithMargin(bounds, otherBounds, margin)) {
          diagnostics.push(diagnostic(rule, "BRAND_LOGO_CLEAR_SPACE_VIOLATION", node.id, margin, other.id));
          break;
        }
      }
    }
  }
  return diagnostics;
}

function evaluateBinding(rule: BrandRule, node: DesignNode, context: BrandEvaluationContext): BrandDiagnostic[] {
  if (rule.type !== "REQUIRE_TOKEN_BINDING") return [];
  const binding = nodeBrandBinding(node);
  const prefixes = stringArray(rule.parameters.prefixes);
  if (binding && (!prefixes.length || prefixes.some((prefix) => binding.startsWith(prefix)))) return [];
  const suggested = prefixes[0];
  return [diagnostic(
    rule,
    "BRAND_TOKEN_BINDING_REQUIRED",
    node.id,
    prefixes,
    binding ?? null,
    suggested ? [repairOperation(context, node.id, "brand_binding", suggested, rule.id)] : undefined,
  )];
}

function evaluateAsset(rule: BrandRule, node: DesignNode, context: BrandEvaluationContext): BrandDiagnostic[] {
  if (rule.type !== "ALLOWED_ASSETS" || !["IMAGE", "VIDEO", "VECTOR_PATH"].includes(node.kind)) return [];
  const assetId = nodeAssetId(node);
  const allowed = new Set(stringArray(rule.parameters.asset_ids));
  if (!assetId || !allowed.has(assetId)) return [diagnostic(rule, "BRAND_ASSET_NOT_ALLOWED", node.id, [...allowed], assetId ?? null)];
  if (context.verified_asset_ids && !context.verified_asset_ids.includes(assetId)) return [diagnostic(rule, "BRAND_ASSET_NOT_VERIFIED", node.id, true, assetId)];
  return [];
}

function evaluateVoice(rule: BrandRule, node: DesignNode): BrandDiagnostic[] {
  if (node.kind !== "TEXT") return [];
  const text = nodeText(node);
  if (!text) return [];
  if (rule.type === "VOICE_FORBIDDEN_TERMS") {
    const forbidden = stringArray(rule.parameters.terms);
    const lowered = text.toLocaleLowerCase();
    const hit = forbidden.find((term) => lowered.includes(term.toLocaleLowerCase()));
    return hit ? [diagnostic(rule, "BRAND_VOICE_FORBIDDEN_TERM", node.id, forbidden, hit)] : [];
  }
  return [];
}

function evaluateRule(rule: BrandRule, context: BrandEvaluationContext): BrandDiagnostic[] {
  if (!rule.active) return [];
  const diagnostics: BrandDiagnostic[] = [];
  const nodes = Object.values(context.document.nodes).sort((a, b) => a.id.localeCompare(b.id));
  for (const node of nodes) {
    if (!appliesToRule(rule, node, context)) continue;
    diagnostics.push(...evaluateColor(rule, node, context));
    diagnostics.push(...evaluateTypography(rule, node, context));
    diagnostics.push(...evaluateLogo(rule, node, context));
    diagnostics.push(...evaluateBinding(rule, node, context));
    diagnostics.push(...evaluateAsset(rule, node, context));
    diagnostics.push(...evaluateVoice(rule, node));
  }
  return diagnostics;
}

export function validateBrandRuleSet(ruleSet: BrandRuleSet): void {
  if (!ruleSet.id || !ruleSet.brand_profile_id || !ruleSet.version) throw new BrandRuleError("brand rule set identity is required");
  if (new Set(ruleSet.rules.map((rule) => rule.id)).size !== ruleSet.rules.length) throw new BrandRuleError("brand rule ids must be unique");
  for (const rule of ruleSet.rules) {
    if (!Number.isFinite(rule.priority)) throw new BrandRuleError(`rule ${rule.id} priority must be finite`);
    if (rule.source === "INFERRED_PROPOSAL" && rule.severity === "HARD") {
      throw new BrandRuleError(`inferred proposal ${rule.id} cannot be HARD`);
    }
    if (rule.source === "APPROVED_GUIDE_EXTRACTION" && !rule.citations?.length) {
      throw new BrandRuleError(`approved guide rule ${rule.id} requires source citations`);
    }
  }
  if (ruleSet.status === "PUBLISHED" && ruleSet.rules.some((rule) => rule.source === "INFERRED_PROPOSAL")) {
    throw new BrandRuleError("published rule sets cannot contain unreviewed inferred proposals");
  }
}

export function publishBrandRuleSet(ruleSet: BrandRuleSet, publishedAt: string): BrandRuleSet {
  const candidate: BrandRuleSet = { ...ruleSet, status: "PUBLISHED", published_at: publishedAt };
  validateBrandRuleSet(candidate);
  return candidate;
}

export function evaluateBrandCompliance(context: BrandEvaluationContext): BrandComplianceReport {
  validateBrandRuleSet(context.rule_set);
  if (context.rule_set.status !== "PUBLISHED") throw new BrandRuleError("compliance requires a PUBLISHED BrandRuleSet");
  if (context.token_set.version !== context.rule_set.token_set_version) throw new BrandRuleError("token set version mismatch");
  if (context.asset_set.version !== context.rule_set.asset_set_version) throw new BrandRuleError("asset set version mismatch");

  const diagnostics = context.rule_set.rules
    .slice()
    .sort((a, b) => b.priority - a.priority || a.id.localeCompare(b.id))
    .flatMap((rule) => evaluateRule(rule, context));
  const hard = diagnostics.filter((item) => item.severity === "HARD").length;
  const soft = diagnostics.filter((item) => item.severity === "SOFT").length;
  const advisory = diagnostics.filter((item) => item.severity === "ADVISORY").length;
  const penalty = hard * 0.25 + soft * 0.08 + advisory * 0.02;
  return {
    brand_rule_set_version: context.rule_set.version,
    decision: hard > 0 ? "FAIL" : diagnostics.length > 0 ? "PASS_WITH_WARNINGS" : "PASS",
    score: Math.max(0, Math.min(1, 1 - penalty)),
    diagnostics,
    hard_violation_count: hard,
    soft_violation_count: soft,
    advisory_count: advisory,
  };
}
